Geometry_Engine/Modify/Extend.cs

```diff
@@ -21,22 +21,355 @@
+            {
+                double endCut = -curves[curves.Count - 1].ILength();
+                while (endCut > end)
+                {
+                    curves.RemoveAt(curves.Count - 1);
+                    endCut -= curves[curves.Count - 1].ILength();
+                }
+                endCut += curves[curves.Count - 1].ILength();
+                curves[curves.Count - 1] = curves[curves.Count - 1].IExtend(0, end - endCut, tangentExtensions, tolerance);
+            }
+            else
+                curves[curves.Count - 1] = curves[curves.Count - 1].IExtend(0, end, tangentExtensions, tolerance);
+
+            return new PolyCurve { Curves = curves };
+        }
+
+
+        /***************************************************/
+        /***   Public Methods - Interfaces               ***/
+        /***************************************************/
+
+        [Description("Extends curve by given lengths")]
+        [Input("curve", "Curve to extend")]
+        [Input("start", "Length of extension on the begining of a curve. Negative value will trim the curve")]
+        [Input("end", "Length of extension on the end of a curve. Negative value will trim the curve")]
+        [Input("tangentExtensions", "True - extends on tangent lines\nFalse - extends by curve's shape")]
+        [Output("curve", "Extended curve")]
+        public static ICurve IExtend(this ICurve curve, double start = 0.0, double end = 0.0, bool tangentExtensions = false, double tolerance = Tolerance.Distance)
+        {
+            return Extend(curve as dynamic, start, end, tangentExtensions, tolerance);
+        }
+
+
+        /***************************************************/
+        /***   Private Methods - Tangent extensions      ***/
+        /***************************************************/
+
+        private static ICurve ExtendTangent(this Arc curve, double start = 0.0, double end = 0.0, double tolerance = Tolerance.Distance)
+        {
+            if (-start > curve.Length() || -end > curve.Length()) //ExtendTangent allows to trim the curve but only to the limit of it's length. 
+            {
+                Reflection.Compute.RecordError("Extension value to small");
+                return null;
+            }
+            if (start < 0 && end < 0)
+                return curve.Extend(start, end, false, tolerance);
+
+            PolyCurve polyCurve = new PolyCurve();
+
+            if (start < 0 && !(end < 0))
+            {
+                ICurve iCurve = curve.Extend(start, 0, false, tolerance);
+                return iCurve.IExtend(0, end, true, tolerance);
+            }
+
+            if (!(start < 0) && end < 0)
+            {
+                ICurve iCurve = curve.Extend(0, end, false, tolerance);
+                return iCurve.IExtend(start, 0, true, tolerance);
+            }
+
+            Point stPt = curve.StartPoint();
+            Point enPt = curve.EndPoint();
+            Vector startTan = curve.TangentAtParameter(0);
```

Comment:
`TangentAtParameter` should pass `tolerance` (please double check other methods too)